packages/block-editor/src/store/actions.js

```diff
@@ -37,19 +37,48 @@ function* ensureDefaultBlock() {
 }
 
 /**
- * Returns an action object used in signalling that blocks state should be
- * reset to the specified array of blocks, taking precedence over any other
- * content reflected as an edit in state.
+ * Generator which resets blocks, also resetting selection state if required.
  *
  * @param {Array} blocks Array of blocks.
- *
- * @return {Object} Action object.
  */
-export function resetBlocks( blocks ) {
-	return {
+export function* resetBlocks( blocks ) {
+	yield {
 		type: 'RESET_BLOCKS',
 		blocks,
 	};
+
+	const selectionStartClientId = yield select(
+		'core/block-editor',
+		'getBlockSelectionStart'
+	);
+	const selectionEndClientId = yield select(
+		'core/block-editor',
+		'getBlockSelectionEnd'
+	);
+
+	// Return early if there are no selected blocks, nothing further needs to
+	// be done.
+	if ( ! selectionStartClientId && ! selectionEndClientId ) {
+		return;
+	}
+
+	const selectionStartBlock = yield select(
+		'core/block-editor',
+		'getBlock',
+		selectionStartClientId
+	);
+
+	const selectionEndBlock = yield select(
+		'core/block-editor',
+		'getBlock',
+		selectionEndClientId
+	);
+
+	// If there was a selected block, but it no longer exists after
+	// RESET_BLOCKS, clear the block selection state.
+	if ( ! selectionStartBlock || ! selectionEndBlock ) {
+		yield clearSelectedBlock();
+	}
```

Comment:
Do you think it'd be better to solve this with a Higher-order reducer instead to avoid two dispatches?

Reply:
@youknowriad Initially looked into making it part of the reducer using the `selection` helper that's already there (https://github.com/WordPress/gutenberg/pull/21371), but the trouble I had is that `selectionStart` and `selectionEnd` are two separate reducers, but the code needs to access both bits of state to determine if selection should be cleared.

Not sure if that's something that can be solved at reducer level without combining the two reducers.

Reply:
@talldan yes it can we have a few instances of this `withBlockReset` is one example.

Reply:
Ah, ok great, will take a look 👍